File: Kotlin-Sample/app/src/main/java/com/freestar/sample/kotlin/MediationPartners.kt
package com.freestar.sample.kotlin

import android.app.AlertDialog
import android.content.Context
import android.content.DialogInterface
import com.freestar.android.ads.AdRequest
import com.freestar.android.ads.LVDOConstants.PARTNER
import java.util.Collections

/**
 * Only for testing purposes!
 */
object MediationPartners {
    const val ADTYPE_INTERSTITIAL = 0
    const val ADTYPE_REWARDED = 1
    const val ADTYPE_BANNER = 2
    const val ADTYPE_PREROLL = 3

    /**
     * INTERSTITIAL
     */
    private val sInterstitialPartners: MutableList<String> = ArrayList(30)
    private val sSelectedInterstitialPartners: BooleanArray

    init {
        sInterstitialPartners.add(PARTNER.ADCOLONY.name)
        sInterstitialPartners.add(PARTNER.APPLOVIN.name)
        sInterstitialPartners.add(PARTNER.APPLOVINMAX.name)
        sInterstitialPartners.add(PARTNER.CRITEO.name)
        sInterstitialPartners.add(PARTNER.GOOGLEADMOB.name)
        sInterstitialPartners.add(PARTNER.GOOGLE.name)
        sInterstitialPartners.add(PARTNER.HYPRMX.name)
        sInterstitialPartners.add(PARTNER.NIMBUS.name)
        sInterstitialPartners.add(PARTNER.PANGLE.name)
        sInterstitialPartners.add(PARTNER.TAM.name)
        sInterstitialPartners.add(PARTNER.TAPJOY.name)
        sInterstitialPartners.add(PARTNER.UNITY.name)
        sInterstitialPartners.add(PARTNER.VUNGLE.name)
        sInterstitialPartners.add(PARTNER.YAHOO.name)
        sInterstitialPartners.add(PARTNER.PREBID.name)
        sInterstitialPartners.add(PARTNER.OGURY.name)
        sInterstitialPartners.add(PARTNER.FYBER.name)
        sInterstitialPartners.add(PARTNER.SMAATO.name)
        Collections.sort(sInterstitialPartners)
        sSelectedInterstitialPartners = BooleanArray(sInterstitialPartners.size)
        for (i in sInterstitialPartners.indices) {
            sSelectedInterstitialPartners[i] = true
        }
    }

    private fun setInterstitialPartners(adRequest: AdRequest): List<PARTNER> {
        val list: MutableList<PARTNER> = ArrayList(sInterstitialPartners.size)
        for (i in sInterstitialPartners.indices) {
            if (sSelectedInterstitialPartners[i]) {
                list.add(PARTNER.valueOf(sInterstitialPartners[i]))
            }
        }
        adRequest.partnerNames = list
        return list
    }

    /**
     * REWARDED
     */
    private val sRewardedPartners: MutableList<String> = ArrayList(30)
    private val sSelectedRewardedPartners: BooleanArray

    init {
        sRewardedPartners.add(PARTNER.ADCOLONY.name)
        sRewardedPartners.add(PARTNER.APPLOVIN.name)
        sRewardedPartners.add(PARTNER.APPLOVINMAX.name)
        sRewardedPartners.add(PARTNER.CRITEO.name)
        sRewardedPartners.add(PARTNER.GOOGLEADMOB.name)
        sRewardedPartners.add(PARTNER.GOOGLE.name)
        sRewardedPartners.add(PARTNER.NIMBUS.name)
        sRewardedPartners.add(PARTNER.TAPJOY.name)
        sRewardedPartners.add(PARTNER.UNITY.name)
        sRewardedPartners.add(PARTNER.VUNGLE.name)
        sRewardedPartners.add(PARTNER.PANGLE.name)
        sRewardedPartners.add(PARTNER.HYPRMX.name)
        sRewardedPartners.add(PARTNER.PREBID.name)
        sRewardedPartners.add(PARTNER.OGURY.name)
        sRewardedPartners.add(PARTNER.FYBER.name)
        sRewardedPartners.add(PARTNER.SMAATO.name)
        Collections.sort(sRewardedPartners)
        sSelectedRewardedPartners = BooleanArray(sRewardedPartners.size)
        for (i in sRewardedPartners.indices) {
            sSelectedRewardedPartners[i] = true
        }
    }

    private fun setRewardedPartners(adRequest: AdRequest): List<PARTNER> {
        val list: MutableList<PARTNER> = ArrayList(sRewardedPartners.size)
        for (i in sRewardedPartners.indices) {
            if (sSelectedRewardedPartners[i]) {
                list.add(PARTNER.valueOf(sRewardedPartners[i]))
            }
        }
        adRequest.partnerNames = list
        return list
    }

    /**
     * NATIVE INVIEW
     */
    private val sBannerPartners: MutableList<String> = ArrayList(30)
    private val sSelectedBannerPartners: BooleanArray

    init {
        sBannerPartners.add(PARTNER.TAM.name)
        sBannerPartners.add(PARTNER.ADCOLONY.name)
        sBannerPartners.add(PARTNER.APPLOVIN.name)
        sBannerPartners.add(PARTNER.APPLOVINMAX.name)
        sBannerPartners.add(PARTNER.CRITEO.name)
        sBannerPartners.add(PARTNER.GOOGLEADMOB.name)
        sBannerPartners.add(PARTNER.GOOGLE.name)
        sBannerPartners.add(PARTNER.NIMBUS.name)
        sBannerPartners.add(PARTNER.UNITY.name)
        sBannerPartners.add(PARTNER.PANGLE.name)
        sBannerPartners.add(PARTNER.VUNGLE.name)
        sBannerPartners.add(PARTNER.HYPRMX.name)
        sBannerPartners.add(PARTNER.YAHOO.name)
        sBannerPartners.add(PARTNER.PREBID.name)
        sBannerPartners.add(PARTNER.OGURY.name)
        sBannerPartners.add(PARTNER.FYBER.name)
        sBannerPartners.add(PARTNER.SMAATO.name)
        Collections.sort(sBannerPartners)
        sSelectedBannerPartners = BooleanArray(sBannerPartners.size)
        for (i in sBannerPartners.indices) {
            sSelectedBannerPartners[i] = true
        }
    }

    private fun setInviewPartners(adRequest: AdRequest): List<PARTNER> {
        val list: MutableList<PARTNER> = ArrayList(sBannerPartners.size)
        for (i in sBannerPartners.indices) {
            if (sSelectedBannerPartners[i]) {
                list.add(PARTNER.valueOf(sBannerPartners[i]))
            }
        }
        adRequest.partnerNames = list
        return list
    }

    /**
     * PRE-ROLL
     */
    private val sPrerollPartners: MutableList<String> = ArrayList(30)
    private val sSelectedPrerollPartners: BooleanArray

    init {
        sPrerollPartners.add(PARTNER.GOOGLE.name)
        sPrerollPartners.add(PARTNER.NIMBUS.name)
        Collections.sort(sPrerollPartners)
        sSelectedPrerollPartners = BooleanArray(sPrerollPartners.size)
        for (i in sPrerollPartners.indices) {
            sSelectedPrerollPartners[i] = true
        }
    }

    private fun setPrerollPartners(adRequest: AdRequest): List<PARTNER> {
        val list: MutableList<PARTNER> = ArrayList(sPrerollPartners.size)
        for (i in sPrerollPartners.indices) {
            if (sSelectedPrerollPartners[i]) {
                list.add(PARTNER.valueOf(sPrerollPartners[i]))
            }
        }
        adRequest.partnerNames = list
        return list
    }

    /**
     * @param adUnitType
     * @param context
     * @param listener
     */
    fun choosePartners(
        context: Context?,
        adRequest: AdRequest,
        adUnitType: Int,
        listener: DialogInterface.OnClickListener
    ) {
        val partners: Array<String?>
        val selected: BooleanArray
        val title: String
        if (adUnitType == ADTYPE_INTERSTITIAL) {
            //partners = arrayOfNulls(sInterstitialPartners.size)
            //sInterstitialPartners.toArray<String>(partners)
            partners = sInterstitialPartners.toTypedArray()
            selected = sSelectedInterstitialPartners
            title = "Interstitial"
        } else if (adUnitType == ADTYPE_REWARDED) {
            partners = sRewardedPartners.toTypedArray()
            selected = sSelectedRewardedPartners
            title = "Rewarded"
        } else if (adUnitType == ADTYPE_BANNER) {
            partners = sBannerPartners.toTypedArray()
            selected = sSelectedBannerPartners
            title = "Display"
        } else {
            partners = sPrerollPartners.toTypedArray()
            selected = sSelectedPrerollPartners
            title = "Pre-Roll"
        }
        //w/out the listener, even though it does nothing, things don't work so well
        //so we need to pass a listener
        AlertDialog.Builder(context)
            .setMultiChoiceItems(partners, selected) { dialog, which, isChecked -> }
            .setPositiveButton("OK") { dialog, which ->
                when (adUnitType) {
                    ADTYPE_INTERSTITIAL -> setInterstitialPartners(adRequest)
                    ADTYPE_BANNER -> setInviewPartners(adRequest)
                    ADTYPE_PREROLL -> setPrerollPartners(adRequest)
                    ADTYPE_REWARDED -> setRewardedPartners(adRequest)
                    else -> {}
                }
                listener.onClick(dialog, which)
            }
            .setNegativeButton("CANCEL", dummyOnClick)
            .setTitle(title)
            .show()
    }

    private val dummyOnClick = DialogInterface.OnClickListener { dialog, which -> }

    /**
     * @param adUnitType 0:interstitial, 1:rewarded, 2:inview, 3:preroll
     */
    fun getChosenPartners(adUnitType: Int): Array<String?> {
        val partners: Array<String?>
        val selected: BooleanArray
        val chosen: Array<String?>
        val total: Int
        if (adUnitType == ADTYPE_INTERSTITIAL) {
            partners = sInterstitialPartners.toTypedArray()
            selected = sSelectedInterstitialPartners
            total = sInterstitialPartners.size
            chosen = arrayOfNulls(sInterstitialPartners.size)
        } else if (adUnitType == ADTYPE_REWARDED) {
            partners = sRewardedPartners.toTypedArray()
            selected = sSelectedRewardedPartners
            total = sRewardedPartners.size
            chosen = arrayOfNulls(sRewardedPartners.size)
        } else if (adUnitType == ADTYPE_BANNER) {
            partners = sBannerPartners.toTypedArray()
            selected = sSelectedBannerPartners
            total = sBannerPartners.size
            chosen = arrayOfNulls(sBannerPartners.size)
        } else {
            partners = sPrerollPartners.toTypedArray()
            selected = sSelectedPrerollPartners
            total = sPrerollPartners.size
            chosen = arrayOfNulls(sPrerollPartners.size)
        }
        var j = 0
        for (i in 0 until total) {
            if (selected[i]) {
                chosen[j++] = partners[i]
            }
        }
        return chosen
    }
}

File: Xamarin-Android-Sample/MediationPartners.cs
using Android.Content;
using Com.Freestar.Android.Ads;

namespace fs_sample
{
    public class MediationPartners
    {
        private static int NUM_INTERSTITIAL = 16;
        private static int NUM_REWARDED = 14;
        private static int NUM_BANNER = 15;

        private static string[] rewarded_partners = new string[NUM_REWARDED];
        private static string[] interstitial_partners = new string[NUM_INTERSTITIAL];
        private static string[] banner_partners = new string[NUM_BANNER];

        private static bool[] rewardedSelected = new bool[NUM_REWARDED];
        private static bool[] interstitialSelected = new bool[NUM_INTERSTITIAL];
        private static bool[] bannerSelected = new bool[NUM_BANNER];

        private static bool isInitialized;

        public static void init()
        {
            rewarded_partners[0] = LVDOConstants.PARTNER.Adcolony.Name();
            rewarded_partners[1] = LVDOConstants.PARTNER.Applovin.Name();
            rewarded_partners[2] = LVDOConstants.PARTNER.Applovinmax.Name();
            rewarded_partners[3] = LVDOConstants.PARTNER.Criteo.Name();
            rewarded_partners[4] = LVDOConstants.PARTNER.Facebook.Name();
            rewarded_partners[5] = LVDOConstants.PARTNER.Googleadmob.Name();
            rewarded_partners[6] = LVDOConstants.PARTNER.Google.Name();
            rewarded_partners[7] = LVDOConstants.PARTNER.Nimbus.Name();
            rewarded_partners[8] = LVDOConstants.PARTNER.Tapjoy.Name();
            rewarded_partners[9] = LVDOConstants.PARTNER.Unity.Name();
            rewarded_partners[10] = LVDOConstants.PARTNER.Vungle.Name();
            rewarded_partners[11] = LVDOConstants.PARTNER.Pangle.Name();
            rewarded_partners[12] = LVDOConstants.PARTNER.Hyprmx.Name();
            rewarded_partners[13] = LVDOConstants.PARTNER.Prebid.Name();

            interstitial_partners[0] = LVDOConstants.PARTNER.Tam.Name();
            interstitial_partners[1] = LVDOConstants.PARTNER.Adcolony.Name();
            interstitial_partners[2] = LVDOConstants.PARTNER.Applovin.Name();
            interstitial_partners[3] = LVDOConstants.PARTNER.Applovinmax.Name();
            interstitial_partners[4] = LVDOConstants.PARTNER.Criteo.Name();
            interstitial_partners[5] = LVDOConstants.PARTNER.Facebook.Name();
            interstitial_partners[6] = LVDOConstants.PARTNER.Googleadmob.Name();
            interstitial_partners[7] = LVDOConstants.PARTNER.Google.Name();
            interstitial_partners[8] = LVDOConstants.PARTNER.Nimbus.Name();
            interstitial_partners[9] = LVDOConstants.PARTNER.Tapjoy.Name();
            interstitial_partners[10] = LVDOConstants.PARTNER.Unity.Name();
            interstitial_partners[11] = LVDOConstants.PARTNER.Vungle.Name();
            interstitial_partners[12] = LVDOConstants.PARTNER.Pangle.Name();
            interstitial_partners[13] = LVDOConstants.PARTNER.Hyprmx.Name();
            interstitial_partners[14] = LVDOConstants.PARTNER.Yahoo.Name();
            interstitial_partners[15] = LVDOConstants.PARTNER.Prebid.Name();

            banner_partners[0] = LVDOConstants.PARTNER.Tam.Name();
            banner_partners[1] = LVDOConstants.PARTNER.Adcolony.Name();
            banner_partners[2] = LVDOConstants.PARTNER.Applovin.Name();
            banner_partners[3] = LVDOConstants.PARTNER.Applovinmax.Name();
            banner_partners[4] = LVDOConstants.PARTNER.Criteo.Name();
            banner_partners[5] = LVDOConstants.PARTNER.Facebook.Name();
            banner_partners[6] = LVDOConstants.PARTNER.Googleadmob.Name();
            banner_partners[7] = LVDOConstants.PARTNER.Google.Name();
            banner_partners[8] = LVDOConstants.PARTNER.Nimbus.Name();
            banner_partners[9] = LVDOConstants.PARTNER.Unity.Name();
            banner_partners[10] = LVDOConstants.PARTNER.Pangle.Name();
            banner_partners[11] = LVDOConstants.PARTNER.Vungle.Name();
            banner_partners[12] = LVDOConstants.PARTNER.Hyprmx.Name();
            banner_partners[13] = LVDOConstants.PARTNER.Yahoo.Name();
            banner_partners[14] = LVDOConstants.PARTNER.Prebid.Name();

            for (int i=0; i < NUM_BANNER;i++)
            {
                bannerSelected[i] = true;
            }
            for (int i=0; i < NUM_REWARDED;i++)
            {
                rewardedSelected[i] = true;
            }
            for (int i = 0; i < NUM_INTERSTITIAL; i++)
            {
                interstitialSelected[i] = true;
            }
        }

        public static void setInterstitialPartners(AdRequest adRequest)
        {
            if (adRequest.PartnerNames != null)
            {
                adRequest.PartnerNames.Clear();
            }
            for (int i=0; i < NUM_INTERSTITIAL;i++)
            {
                if (interstitialSelected[i])
                {
                    adRequest.AddPartnerName(LVDOConstants.PARTNER.ValueOf(interstitial_partners[i]));
                }
            }
        }

        public static void setRewardedPartners(AdRequest adRequest)
        {
            if (adRequest.PartnerNames != null)
            {
                adRequest.PartnerNames.Clear();
            }
            for (int i = 0; i < NUM_REWARDED; i++)
            {
                if (rewardedSelected[i])
                {
                    adRequest.AddPartnerName(LVDOConstants.PARTNER.ValueOf(rewarded_partners[i]));
                }
            }
        }

        public static void setBannerPartners(AdRequest adRequest)
        {
            if (adRequest.PartnerNames != null)
            {
                adRequest.PartnerNames.Clear();
            }
            for (int i = 0; i < NUM_BANNER; i++)
            {
                if (bannerSelected[i])
                {
                    adRequest.AddPartnerName(LVDOConstants.PARTNER.ValueOf(banner_partners[i]));
                }
            }
        }

        public static void choosePartners(Context context, AdRequest adRequest, int adUnitType, IDialogInterfaceOnClickListener listener)
        {

            if (!isInitialized)
            {
                isInitialized = true;
                init();
            }
            string[] partners;
            bool[] selected;
            string title;
            if (adUnitType == MainActivity.ADTYPE_INTERSTITIAL)
            {
                partners = interstitial_partners;
                selected = interstitialSelected;
                title = "Interstitial";
            }
            else if (adUnitType == MainActivity.ADTYPE_REWARDED)
            {
                partners = rewarded_partners;
                selected = rewardedSelected;
                title = "Rewarded";
            }
            else if (adUnitType == MainActivity.ADTYPE_BANNER)
            {
                partners = banner_partners;
                selected = bannerSelected;
                title = "Banner";
            } else
            {
                partners = banner_partners;
                selected = bannerSelected;
                title = "MREC";
            }
            Android.App.AlertDialog.Builder dialog = new Android.App.AlertDialog.Builder(context);
            OnMultiClickListener multiListener = new OnMultiClickListener(selected);
            dialog.SetTitle(title);
            dialog.SetMultiChoiceItems(partners, selected, multiListener);
            dialog.SetNegativeButton("CANCEL", new DummyListener());
            dialog.SetPositiveButton("OK", new OnOkClickListener(adUnitType, adRequest, listener));
            dialog.Show();
        }

    }

}